### PR TITLE
Add block cache for low level table library

### PR DESCRIPTION
**Pull request by byronyi**
This is part of a patch series aiming to improve the performance of on-disk dataset.cache() (CacheDatasetV2).

Currently CacheDataset uses core/util/tensor_bundle to cache dataset elements on disks. It uses sorted string table (SST) to index dataset elements. Unlike checkpoints which do not have a great number of tensors, caching a large dataset may incur a greater number of tensors as well as index blocks.

If the index block is present in an in-memory LRU block cache, fetching a dataset element only needs 1 round trip instead of 2. This is particularly useful when CacheDataset are read from remote file system at a higher latency such as HDFS and GCS.

Almost all code are imported from the LevelDB project, in particular the hash function to shard LRU cache. Currently using Hash32 in core/lib/hash fails the EvictionPolicy test.

I only make 2 modifications to the original cache:

1. Alias leveldb::Slice to tensorflow::StringPiece, which transitively aliases to absl::string_view.
2. Switch to tensorflow::mutex for all mutexes.

Ping @jsimsa to review.

### CODE REVIEW

File: tensorflow/core/lib/io/cache_test.cc [hunk 1]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
Comment (jsimsa) [inline tensorflow/core/lib/io/cache_test.cc]: this should be using the TensorFlow copyright

File: tensorflow/core/lib/io/cache.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
Comment (jsimsa) [inline tensorflow/core/lib/io/cache.h]: this should be using the TensorFlow copyright

File: tensorflow/core/lib/io/cache.cc [hunk 3]
```diff
@@ -0,0 +1,439 @@
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
Comment (jsimsa) [inline tensorflow/core/lib/io/cache.cc]: this should be using the TensorFlow copyright

File: tensorflow/core/lib/io/cache_test.cc (hunk 1 above)
Comment (byronyi) [reply tensorflow/core/lib/io/cache_test.cc]: Changed.

File: tensorflow/core/lib/io/cache.h (hunk 2 above)
Comment (byronyi) [reply tensorflow/core/lib/io/cache.h]: Changed.

File: tensorflow/core/lib/io/cache.cc (hunk 3 above)
Comment (byronyi) [reply tensorflow/core/lib/io/cache.cc]: Changed.

**Comment by byronyi** [comment]
Gently ping @gbaned for reviewing. 

It seems we have r2.2 branch cut later this month and I would really like this PR to make it into the next release.

**Review by jsimsa — DISMISSED** [review]

**Comment by byronyi** [comment]
@gbaned Hi, is there anything I could do to roll this PR forward?

**Comment by gbaned** [comment]
> @gbaned Hi, is there anything I could do to roll this PR forward?

@byronyi Sorry for the slow response, it is processing internally. We will let you know if anything needed. Thank you.

File: tensorflow/core/lib/io/cache_test.cc [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/lib/io/cache.h"
+
+#include <string>
+#include <vector>
+
+#include "tensorflow/core/platform/coding.h"
```
Comment (rthadur) [inline tensorflow/core/lib/io/cache_test.cc]: `third_party/tensorflow/core/lib/io/cache_test.cc:21:10: error: use of private header from outside its module: 'third_party/tensorflow/core/platform/coding.h'; `

**Review by rthadur — CHANGES_REQUESTED** [review]

File: tensorflow/core/lib/io/cache_test.cc (hunk 4 above)
Comment (rthadur) [reply tensorflow/core/lib/io/cache_test.cc]: @byronyi can you please check this error ?

Comment (byronyi) [reply tensorflow/core/lib/io/cache_test.cc]: @rthadur Fixed. Mind to kickoff CI again?

**Comment by byronyi** [comment]
@gbaned I've fixed the review comment. Mind to kick off CI tests again?

**Review by rthadur — APPROVED** [review]

**Comment by byronyi** [comment]
@gbaned mind to take a look? r2.2 branch cut draws near.

**Comment by byronyi** [comment]
I see this is being merged manually. Thanks!

**Pull request closed without merging**